### DISCUSSION TITLE
"PCB Exposure" issue

### DISCUSSION THREAD

**Discussion by ghost** [sn4k3/UVtools#492] [Q&A]
Hi all, I am starting with my resin 3D printer and I have discovered this "UVTools" which is perfect for me to optimize normal 3D printing except when I try to convert a Gerber file using the "PCB Exposure" tool getting this weird Message :
![Capture](https://user-images.githubusercontent.com/107458958/173507336-48160fce-2cf2-4332-a6be-e4d7a4258b65.JPG)

The same Gerber file worked flawlessly when I sent it to my favorite PCB service.

The strange thing is the path "D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.Core\...." which is not present in my drive D: ...

Anyone get the same error or fixed it?

Thanks to those who want to answer me.

**Comment by sn4k3** [comment]
PCB tool is still experimental, not all gerber functions are implemented, for example arcs and inch meassure
I can't debug the problem just with your message, i will need the gerber file

**Comment by ghost** [comment]
Thanks for your reply, I will send everything necessary to reproduce the
problem as soon as possible. Furthermore there are also some problems with
a normal use of UVTools, eg. when uploading a PNG image to create a sliced
file for my Anycubic 3D printer.
I will get in touch quickly.
Greetings

**Reply by sn4k3** [reply]
Check with latest version, does it still happens?

**Comment by ghost** [comment]
Thanks for the reply ... yes, now it works except that there are no holes
inside the pads and no possibility to load excellon/drill files to "drill"
them (see attached snapshot gbr_test_1.jpg) BUT if I were to export it same
layer together with the PAD layer as PDF via the EagleCAD "Print to PDF
file" function, then the created PDF file is imported into the GerbView app
(which loads the PDF files) and exported as a Gerber .gbr file, now the
latter in the UVTools preview will be displayed with an inconclusive image
(see attached snapshot gbr_test_2.jpg). The gbr file is not damaged,
ZoftPCB 3D gerber Viewer loads it correctly (see zoftpcb.jpg). In the end,
of course, I attach both gbr files for checking (note that the square
inside the PCB is one inch per side and the EXPORT LAYERS to MESH-STL does
not work at all, it generate a tangle of intertwined polygons).

Just a question: Wouldn't it have been easier to implement importing pdf
files with UVTools? it could open up a world of possibilities to use other
"PDF saving/export" software such as Corel draw, Inkscape, ecc. for
creating stunning 3D panels from 2D works.

I will donate some money in the future for supporting you.
![gbr_test_1](https://user-images.githubusercontent.com/107458958/174492325-809f1c80-9791-4431-99a4-4141caf8413d.JPG)
![gbr_test_2](https://user-images.githubusercontent.com/107458958/174492329-4f5fa59b-e902-429b-ba17-081e30797992.JPG)
![zofzPCB](https://user-images.githubusercontent.com/107458958/174492354-900fe9ac-8a10-4f16-9d02-b1dc64d9dd5e.JPG)
[gbr.zip](https://github.com/sn4k3/UVtools/files/8935982/gbr.zip)


Cheers.

**Reply by sn4k3** [reply]
Attachment files via email don't show here.
Please reply directly via github in order to attach them

> Just a question: Wouldn't it have been easier to implement importing pdf files with UVTools? it could open up a world of possibilities to use other "PDF saving/export" software such as Corel draw, Inkscape, ecc. for creating stunning 3D panels from 2D works.

Nothing about PDF is easy. Once converted and even if still a vector, all information is lost and you end with a image which i can't tell what is what.
It's much more easy to interpret source (gerber) and the correct way to implement this stuff, at least all information is kept there.
It's normal my implementation still lacking something, but this kind of reports is what make things keep going and clear bugs.

**Reply by ghost** [reply]
thanks for the reply, I attached jpg and zipped gbr files as stated. Hope they can help you.

**Reply by sn4k3** [reply]
The problem is that the untitled.gbr is in inch as document unit, and i haven't implemented Inch, only millimeters. Can you convert to mm and retry?

**Reply by ghost** [reply]
![from_PDF_gbr_metric](https://user-images.githubusercontent.com/107458958/174495835-2978d198-c1f1-4fe3-9d12-695c1ff3e3f6.JPG)
[untitled.zip](https://github.com/sn4k3/UVtools/files/8936179/untitled.zip)

Converted in mm...Still problems...

**Reply by sn4k3** [reply]
I've implemented the Inch parsing, the result come out better but not perfect. And your mm convertion generate even a worse output.
In my debug i notice that some lines are not in correct order to form the polygon like the TEXT. I'm not sure how gerber viewers handle it, but it seens wrong since polygons are points connected by order, if order is messed it will generate another form.
What kind of CAD software you used to generate the gerbers?

Here the result of the inch file:
![UVtools_2022-06-24_04-44-34](https://user-images.githubusercontent.com/113281/175458005-4c9b6301-ddbe-4a47-aa8d-f7d84a90b6b8.png)

To debug this better, you can try to leave only the TEXT on top and delete everything else, and send me the file so i can have a file with just one error, easier to cath the problem

**Comment by ghost** [comment]
Thanks for your reply and support. As stated in a previous comment, I use EagleCAD plus Gerbview as a PDF to GBR converter.
I have attached two GBR files, see their headers to know which program generated them. Both GBR files produce the same result on any online gerber viewer and even if I import them back into EagleCAD they both work (no errors or messed up traces).

If I try to do the same in UV Tools, the GBR file generated by EagleCAD works, the other does not.

I hope it can help you.

[gerber_TEXT_test.zip](https://github.com/sn4k3/UVtools/files/8973314/gerber_TEXT_test.zip)

**Reply by sn4k3** [reply]
Was able to find the problem, that gerber is using a different zero suppresion mode, after implementation it works well now:

![image](https://user-images.githubusercontent.com/113281/175547163-70361d13-90c8-4f12-beb1-a8796e2b6fe1.png)

Will go on next patch. Thanks for the files.

**Reply by ghost** [reply]
Great, I look forward 👍

**Reply by ghost** [reply]
Ok, I just downloaded v.3.5.2 and it seems to work. Below there is a snapshot of  an example of a PDF created by Corel Draw then converted in GBR by Gerbview and at last imported in UVTools...and it looks perfect. I want just to verify XYZ measures but I should have to calibrate my 3d printer in any case. I sent some money with the account named aviotronic for the support.

![grab](https://user-images.githubusercontent.com/107458958/175782515-b08b2dd1-f487-4368-807a-93f1f4d0ad67.JPG)

**Reply by sn4k3** [reply]
Glad to see it work so well 🙂UVtools is amazing 😉
With ROI tool you can meassure pixels in mm, it should be pretty accurate with an error of 1 or 2 maximum px due no half pixels. But in end it will also depend on your machine and print settings.
Thanks for support!

**Reply by ghost** [reply]
I am still convinced that a "PDF Direct Import Tool" is a better solution and that it is the user who decides to use the result obtained. 
Have you also thought about implementing the "Pixel Shift Multi Shot Technique" to improve the print resolution? It is used in some modern cameras that take four images by moving the image sensor one pixel at a time. You can generate images with a higher resolution than is possible with normal shooting by somehow combining the four images together. This way on every single layer you will perhaps overcome the limitations of the LCD resolution.

**Reply by sn4k3** [reply]
Multi-sampling or color blending are eye candy/illusion and only work on most of us, not machines. LCD output pixels and pixels are pixels which in most cases are squares on mono lcds. 
What works with resin is reduce pixel exposure with a more gray tone to create "half-pixels". Still that is not accurate at all and require a lot of try error using all same variables. I'm not sure if that will work with PCBs since i don't know the process but most likely will not and will introduce noise. Another problem is the light bleed... Your best bet to increase detail and tolerances is to get the best resolution printer in pixel micron size (The lowest possible!) But 50um isn't bad either!

**Reply by TheMiNuS** [reply]
I also have PCB import issue with one of my PCB "generated with Kicad v7.0.11".
This one use more complex shapes and the import seems to miss some elements.
Do I miss something in my gerber generation process ?

**Below the Kicad Gerber generation parameters:**
![image](https://github.com/sn4k3/UVtools/assets/58090715/ca1f4ee4-d364-4e4b-b70b-a93a2687e2b7)

I import the PCB in UVTools 3.4.2 with the resolution target of an Elegoo Saturn 2 (with which I get amazing results with simple pcb shapes compared to classic UV Esposition machine using lazer printed masks).

**Below what I should see in UVTools using Kicad's Gerber viewer:**
![image](https://github.com/sn4k3/UVtools/assets/58090715/c6eddac0-d0f2-401e-8b5a-6debc231f22d)


**Below what I have in UVTools:** (it's flipped compared to above but it's normal as it's something I did, the problem is on some shapes generations that seems to be incomplete)
![image](https://github.com/sn4k3/UVtools/assets/58090715/3b8b57bb-3cec-4932-96aa-a8b883a790f6)


**I also put the source gerber file below.**
[PCB_Import _issue.zip](https://github.com/user-attachments/files/15531878/PCB_Import._issue.zip)


If someone can help me to understand the source of this issue it would be amazing. I struggle with that all my sunday tweaking parameters and I did not find any parameter that helped to generate PCB properly.

MiNuS

**Reply by sn4k3** [reply]
> I import the PCB in UVTools 3.4.2

That is way too old. Update to last version and try again.

**Reply by ghost** [reply]
@TheMiNuS It seems a Kicad issue, Gerbview 10 renders it with issues too, see below
![Immagine 2024-06-03 160401](https://github.com/sn4k3/UVtools/assets/107458958/5c4cb25e-3fc0-43d3-a8e0-c7d5f00e3995)

I suggest to upgrade Kicad to 8.0 to see if this problem will be fixed.
Other online gerber viewers gave same bad result.
Cheers

**Reply by TheMiNuS** [reply]
> > I import the PCB in UVTools 3.4.2
> 
> That is way too old. Update to last version and try again.

Sorry it's a typo I'm using 4.3.2 (the last version)

**Comment by thirstyone** [comment]
which settings in kicad plot dialog should I use to get the smoothest results with uvtools? right now I am seeing several issues. see below: the shape of the pads are somewhat distorted plus clearance issues (the same fragment in kicad for reference)
<img width="494" height="154" alt="Снимок экрана 2026-08-19 в 15 56 49" src="https://github.com/user-attachments/assets/fe64505e-4cb0-4e1e-b4c9-ed7046cc4a3b" />
<img width="544" height="423" alt="Снимок экрана 2026-08-19 в 15 57 18" src="https://github.com/user-attachments/assets/7d09c77f-6a6e-44d7-858f-adc913eec756" />
<img width="720" height="450" alt="Снимок экрана 2026-08-19 в 15 57 37" src="https://github.com/user-attachments/assets/1f0a0f84-4799-44c3-9b8d-31d7b2088d36" />

**Reply by sn4k3** [reply]
Distorced view in UVtools is normal, your printer may be non-square pixel, and it will compensate for it, so you will see distorced view on UVtools, but if you exposure on printer with a papper you can confirm the pad geometry are actually correct.

For PCB Exposure the 1:1 pixel square printers are best for the tool, since the compensation is not always pixel perfect to real geometry
